rust/hyperlane-core/src/accumulator/historic.rs

```diff
@@ -0,0 +1,155 @@
+use crate::{
+    accumulator::{hash_concat, merkle::Proof, ZERO_HASHES},
+    H256,
+};
+
+use super::{merkle::MerkleTree, TREE_DEPTH};
+
+impl MerkleTree {
+    /// Merges the compatible nodes from another merkle tree via DFS.
+    ///
+    /// This should only be run on sparse partial trees, as otherwise
+    /// it can consume quite a lot of memory.
+    pub fn merge(self, b: MerkleTree) -> MerkleTree {
+        match self {
+            MerkleTree::Zero(_) => self,
+            MerkleTree::Leaf(_) => {
+                if self.hash().eq(&b.hash()) {
+                    b
+                } else {
+                    self
+                }
+            }
+            MerkleTree::Node(a_hash, ref a_left, ref a_right) => match b {
+                MerkleTree::Leaf(_) => self,
+                MerkleTree::Zero(_) => self,
+                MerkleTree::Node(_, b_left, b_right) => {
+                    let merged_left = (**a_left).clone().merge((*b_left).clone());
+                    let merged_right = (**a_right).clone().merge((*b_right).clone());
+                    let merged_hash = hash_concat(merged_left.hash(), merged_right.hash());
+                    assert_eq!(merged_hash, a_hash);
+                    MerkleTree::Node(a_hash, Box::new(merged_left), Box::new(merged_right))
+                }
+            },
+        }
+    }
+}
+
+impl Proof {
+    /// Return the proof of this index when it was the latest node in the tree
+    pub fn as_latest(&self) -> Proof {
+        // Replace the right nodes with zero hashes
+        let mut modified_path = [H256::zero(); TREE_DEPTH];
+        for i in 0..TREE_DEPTH {
+            let size = self.index >> i;
+            if (size & 1) == 1 {
+                modified_path[i] = self.path[i];
+            } else {
+                modified_path[i] = ZERO_HASHES[i];
+            }
+        }
+
+        Proof {
+            leaf: self.leaf,
+            index: self.index,
+            path: modified_path,
+        }
+    }
+
+    /// Creates a partial merkle tree out of the proof
```

Comment:
Describing what a "partial merkle tree" is would be really nice, as well as its intended use. It took me working through the code and a drawn out tree to confirm what I thought this was doing.

I actually wonder if we should create a wrapper type to disambiguate a "partial merkle tree" from what we so far have been using the `MerkleTree` struct for (we seem to assume that all the first n leaves are present https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/main/rust/hyperlane-core/src/accumulator/merkle.rs#L25). E.g.:

```
struct PartialMerkleTree(MerkleTree);
```
And we could have `merge` only work for `PartialMerkleTree`. The data structure itself would be exactly the same, but by having a different type we could avoid misuse. Curious what you and @mattiecnvr think of this

Reply:
I think this is a great suggestion, merge is not meant to be called on anything other than a partial tree